### PR TITLE
Support ipywidgets v8

### LINKED ISSUE #1032: `AttributeError` on import while using IPython
I'm getting an `AttributeError` on import when using an IPython interpreter. The CPython interpreter imports it just fine. A full reproduction is [here](https://gist.github.com/mattwthompson/b7ba4f2f7680016ad6af5537b013817f), but basically:

```
$ conda create --name nglview-test
$ conda activate nglview-test
$ conda install nglview -c conda-forge -yq
$ python
Python 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:09:13) [Clang 13.0.1 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import nglview
>>> # command+D to exit out
$ ipython
Python 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:09:13) [Clang 13.0.1 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import nglview
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [1], in <cell line: 1>()
----> 1 import nglview

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/__init__.py:4, in <module>
      1 import warnings
      3 # for doc
----> 4 from . import adaptor, datafiles, show, widget
      5 from ._version import get_versions
      6 from .adaptor import *

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/show.py:13, in <module>
      3 from . import datafiles
      4 from .adaptor import (ASEStructure, ASETrajectory, BiopythonStructure,
      5                       FileStructure, HTMDTrajectory, IODataStructure,
      6                       IOTBXStructure, MDAnalysisTrajectory, MDTrajTrajectory,
   (...)
     11                       RdkitStructure,
     12                       TextStructure)
---> 13 from .widget import NGLWidget
     15 __all__ = [
     16     'demo',
     17     'show_pdbid',
   (...)
     40     'show_biopython',
     41 ]
     44 def show_pdbid(pdbid, **kwargs):

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/widget.py:19, in <module>
     15 from traitlets import (Bool, CaselessStrEnum, Dict, Instance, Int, Integer,
     16                        List, Unicode, observe, validate)
     17 import traitlets
---> 19 from . import color, interpolate
     20 from .adaptor import Structure, Trajectory
     21 from .component import ComponentViewer

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/color.py:114, in <module>
    110         else:
    111             raise ValueError(f"{obj} must be either list of list or string")
--> 114 ColormakerRegistry = _ColormakerRegistry()

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/base.py:10, in _singleton.<locals>.getinstance()
      8 def getinstance():
      9     if cls not in instances:
---> 10         instances[cls] = cls()
     11     return instances[cls]

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/color.py:47, in _ColormakerRegistry.__init__(self, *args, **kwargs)
     45 try:
     46     get_ipython() # only display in notebook
---> 47     self._ipython_display_()
     48 except NameError:
     49     pass

File ~/miniconda3/envs/nglview-test/lib/python3.10/site-packages/nglview/color.py:54, in _ColormakerRegistry._ipython_display_(self, **kwargs)
     52 if self._ready:
     53     return
---> 54 super()._ipython_display_(**kwargs)

AttributeError: 'super' object has no attribute '_ipython_display_'

In [2]:
Do you really want to exit ([y]/n)? y 
```

The [lastest IPython (8.4.0)](https://pypi.org/project/ipython/#history) is from several months ago, so I don't think there'd be a breaking change there. I've been using stuff that imports `nglview` on Python 3.10 for a while, so I doubt it's that. I did notice `ipywidgets` [bumped their major version](https://pypi.org/project/ipython/#history) within the past day (which has been [rolled out](https://github.com/conda-forge/ipywidgets-feedstock/pulls?q=is%3Apr+is%3Aclosed) to `conda-forge` as well) but I don't see that in the traceback.

Any suggestions on how to hunt this down?

- Version report

```
$ python -c 'import nglview; print(nglview.__version__)'
python -c 'import ipywidgets; print(ipywidgets.__version__)'

3.0.3
8.0.1

### PR DESCRIPTION
**Pull request by toslunar**
Fix #1032.

Displaying a controller below the main 3D viewer looks like an exact use case `_ipython_display_`, while `ipywidgets` has replaced `_ipython_display_` with `_repr_mimebundle_`.

[`IPython.core.formatters.IPythonDisplayFormatter`](https://github.com/ipython/ipython/blob/ae0dfcd9d4fb221f0278f07b022f9837b0e51e67/IPython/core/formatters.py#L883-L921) says
> IPython 6.1 introduces `_repr_mimebundle_` for displaying custom mime-types,
> so `_ipython_display_` should only be used for objects that require unusual
> display patterns, such as multiple display calls.

NGLWidget inside another widget still works, while `ipywidgets>=8` seems to assume `_repr_mimebundle_`. It cannot show GUI of NGLWidget if it's inside another widget regardless of the change or the version of ipywidgets.

Another idea to fix the issue is to use `ipywidgets.VBox([viewer, gui])`. It might be more correct fix but
- it would remove the ability to change width of the main viewer; and
- the distinction of the viewer and the box complicates the code.

### CODE REVIEW

**Comment by hainm** [comment]
Dear @toslunar: I've updated master branch to fix the install error. Can you please update `master` branch? cheers.

**Comment by hainm** [comment]
@toslunar  I will fix the tests myself. thanks.

**Comment by toslunar** [comment]
Sorry for my silence. I fixed test failures in my local environment except for `nglview/tests/test_cli.py::test_get_port` which I couldn't reproduce there.

**Comment by toslunar** [comment]
By the way, I'm running the tests with ipywidgets==8.0.4 because `import ipywidgets; ipywidgets.DOMWidget()` fails with 8.0.5. https://github.com/jupyter-widgets/ipywidgets/issues/3731

**Comment by hainm** [comment]
> `nglview/tests/test_cli.py::test_get_port` which I couldn't reproduce there.

`FAILED nglview/tests/test_cli.py::test_get_port - ModuleNotFoundError: No module named 'notebook'`
I guess we need to explicitly add `notebook` as requirement to `.github/nglview-gha.yml`

**Comment by hainm** [comment]
image


```
>       for k, v in embed.Widget.widgets.items():
E       AttributeError: type object 'Widget' has no attribute 'widgets'
```

it seems fine with `ipywidgets 8.0.5`

image

**Comment by hainm** [comment]
> By the way, I'm running the tests with ipywidgets==8.0.4 because `import ipywidgets; ipywidgets.DOMWidget()` fails with 8.0.5. [jupyter-widgets/ipywidgets#3731](https://github.com/jupyter-widgets/ipywidgets/issues/3731)

> .github/nglview-gha.yml

we can add explicitly add `ipykernel` to this file to avoid the failure you quoted.

**Comment by hainm** [comment]
@toslunar 

image

image

File: nglview/widget.py [hunk 1]
```diff
@@ -544,7 +550,14 @@ def display(self, gui=False, style='ngl'):
                 b = HBox([self, self._gui])
                 def on(b):
                     self.handle_resize()
-                b.on_displayed(on)
+                try:
+                    # ipywidgets < 8
+                    b.on_displayed(on)
+                except AttributeError:
+                    logging.warn(
```
Comment (toslunar) [inline nglview/widget.py:557]: I'd use hierarchical logger:
```python
logger = logging.getLogger(__name__)
logger.warn
```

Can I fix Line 53, too?

Comment (hainm) [reply nglview/widget.py:557]: @toslunar yes, please.

**Comment by hainm** [comment]
@toslunar The tests are finally green. We're very close to merging (or do you want to merge it now without the logging change?)

**Review by hainm — APPROVED** [review]

**Comment by mmfarrugia** [comment]
Hello, it looks like this commit fixed the issue with nglview ipywidgets 8 compatibility.  I am sure that I have the latest nglview but am still running into this issue with compatibility with ipywidgets 8.0.4.  To be certain, could you indicate with version of nglview has addressed this issue? 

To be clear, I am referring to the incompatibilities described in issue #1032 , which brought me to this thread.  i have had this issue many times with different packages in the past so I look forward to resolving it in my conda environments with your latest update, thank you!

**Comment by hainm** [comment]
hi @mmfarrugia `nglview 3.0.4` should work with ipywidets 8.

**Comment by mmfarrugia** [comment]
> hi @mmfarrugia `nglview 3.0.4` should work with ipywidets 8.

Thank you, I just triple checked and it looks like it was 3.0.4 but I'm on a distributed network so filenames/pointers for my packages were wonky. All working now, Thank you!

**Pull request merged**